src/main/java/org/terasology/dynamicCities/buildings/BuildingManager.java

```diff
@@ -188,28 +182,47 @@ public void postBegin() {
 
     public Optional<GenericBuildingComponent> getRandomBuildingOfZoneForCulture(String zone, Rect2i shape, CultureComponent cultureComponent) {
         if (buildings.containsKey(zone)) {
-            int max = buildings.get(zone).size();
-            GenericBuildingComponent building;
+            // TODO: needs to be adapted if buildings have a specific spawn chance
+            List<GenericBuildingComponent> availableBuildings =
+                    buildings.get(zone).stream().filter(b -> cultureComponent.availableBuildings.contains(b.name)).collect(Collectors.toList());
```

Comment:
Precompute the list of possible buildings candidates once. Here, the buildings available for `zone` are filtered for those buildings available for the given culture as well. They are stored in a list so that it is simpler to remove elements from it later.